WordPress/gutenberg
# Suggest Pattern transformations that are contextual to the currently selected blocks

## Description
Resolves: https://github.com/WordPress/gutenberg/issues/28736
Part of: https://github.com/WordPress/gutenberg/issues/27575


This is a POC for suggesting block pattern transformations that are contextual to the currently selected blocks.

First of all we need to decide what type of transformations we will suggest/support and we have to keep in mind flows for template parts and any other special cases we can think of. You can find discussions on the above issues and especially [here](https://github.com/WordPress/gutenberg/issues/27575#issuecomment-750293994).

## Rationale behind the current decisions
1. While it makes sense to show patterns for `grouping blocks` (like `Group` or `Columns`) during insertion in `Placeholder` (explored [here](https://github.com/WordPress/gutenberg/pull/29602)), it doesn't make sense to suggest any pattern to transform, as a user could have anything inside as content. So if any selected block falls into this category, do not suggest any transforms.
2. The only exception to the above would be Template Parts (at least from what I thought) and we should have a replacing all contents flow probably.
3. I have skipped any check about transforming selected blocks to available block transforms inside patterns, as I think it would be very confusing to have a really different suggested pattern. An example would be to have a `paragraph` selected and show patterns that do not include `paragraph` but have transformed the `paragraph` to a `quote` block.
4. To show a suggested pattern, **every selected block must have matched with a pattern's block** . This doesn't necessarily mean that the order will be preserved. You can test this in this PR by selecting a paragraph and two headings (in this order) and see the suggested pattern that has the following order: [heading, paragraph, heading]. Also shown at the gif.
5. If a suggested pattern has more blocks than the selected blocks, add them as are from the pattern. Every pattern block that matched with a selected block will be transformed. IMO this is the main reason for doing this and having block patterns in general. We want to inspire/show users alternative designs. 
6. When we have a match with inner blocks like `Buttons`, keep only the blocks from the selected block and skip the inner blocks from the pattern. You can test this with a test `Buttons` pattern I have in this PR.

I have many comments in my code trying to explain as good as I can the flows.




## Testing instructions
1. Test with the demo patterns I created in `lib/test-block-patterns.php` or register one yourselves
2. Create some paragraphs or headings and make different selections (single or multiple) to observe the previews from the selected blocks.
3. Only if all of the selected blocks can be matched in pattern's blocks, it will be shown as an option

## Screenshots 
![patternTransforms](https://user-images.githubusercontent.com/16275880/111307613-5f605d80-8662-11eb-8c81-45d0ddfad046.gif)

## Notes
1. I would like some feedback/thoughts about the handling of patterns transformation first, so design wise focus on a higher level and not details that can be fixed quite easily.
2. There has been some POC work for Template Parts to take into account active variations, but the flow is not complete. We will probably need to create a new Template Part on a selection.

**I would love your thoughts and feedback on this!**

## Checklist:
